pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java

```diff
@@ -125,11 +128,15 @@ public int filterEntriesForConsumer(Optional<MessageMetadata[]> optMetadataArray
             }
             ByteBuf metadataAndPayload = entry.getDataBuffer();
             final int metadataIndex = i + startOffset;
-            final MessageMetadata msgMetadata = optMetadataArray.map(metadataArray -> metadataArray[metadataIndex])
-                    .orElseGet(() -> (entry instanceof EntryAndMetadata)
-                            ? ((EntryAndMetadata) entry).getMetadata()
-                            : Commands.peekAndCopyMessageMetadata(metadataAndPayload, subscription.toString(), -1)
-                    );
+
+            MessageMetadata msgMetadata;
+            if (metadataArray != null) {
+                msgMetadata = metadataArray[metadataIndex];
+            } else if (entry instanceof EntryAndMetadata entryAndMetadata) {
```

Comment:
Don't use the JDK 14 feature in the existing code. cos it will introduce a compatible risk of cherry-picking.

I would like to give this point a `request change` until someone has different options.

/cc @codelipenghui